From: Sebastien Guiriec <s-guiriec@ti.com>
Date: Fri, 11 Feb 2011 12:28:28 +0000 (+0100)
Subject: ASoC: ABE DSP: Add L4 support
X-Git-Url: http://git.omapzoom.org/?p=kernel%2Fomap.git;a=commitdiff_plain;h=58ff83912ba615cba35715eab75359c549cd0751;hp=cba4a1ed3119ff88e13ffeb03dd68a624b5af38d

ASoC: ABE DSP: Add L4 support

Update ABE DSP driver to use L4 direct path for ABE memory
access in ping-pong mode. AESS default address in hwmod is
also changed to use L4 instead of L3.

Change-Id: Ic08e4002b7a337689e72ef26f0e20107398547ac
Signed-off-by: Sebastien Guiriec <s-guiriec@ti.com>
Signed-off-by: Misael Lopez Cruz <misael.lopez@ti.com>
---

Index: kernel/arch/arm/mach-omap2/omap_hwmod_44xx_data.c
===================================================================
--- kernel.orig/arch/arm/mach-omap2/omap_hwmod_44xx_data.c	2011-03-28 12:29:20.308354598 +0900
+++ kernel/arch/arm/mach-omap2/omap_hwmod_44xx_data.c	2011-03-28 15:57:27.708382549 +0900
@@ -652,8 +652,8 @@
 
 static struct omap_hwmod_addr_space omap44xx_aess_addrs[] = {
 	{
-		.pa_start	= 0x49000000,
-		.pa_end		= 0x491f11ff,
+		.pa_start	= 0x40100000,
+		.pa_end		= 0x401f13ff,
 		.flags		= ADDR_TYPE_RT
 	},
 };
Index: kernel/sound/soc/omap/omap-abe-dsp.c
===================================================================
--- kernel.orig/sound/soc/omap/omap-abe-dsp.c	2011-03-28 12:30:01.117144725 +0900
+++ kernel/sound/soc/omap/omap-abe-dsp.c	2011-03-28 15:57:27.708382549 +0900
@@ -1911,8 +1911,7 @@
 				PING_PONG_WITH_MCU_IRQ);
 
 	/* Memory mapping for hw params */
-	runtime->dma_area  = abe->io_base + ABE_DMEM_BASE_OFFSET_MPU +
-				ABE_VM_AESS_OFFSET + dst;
+	runtime->dma_area  = abe->io_base + ABE_DMEM_BASE_OFFSET_MPU + dst;
 	runtime->dma_addr  = 0;
 	runtime->dma_bytes = period_size * 2;
 
@@ -1988,18 +1987,28 @@
 static int aess_mmap(struct snd_pcm_substream *substream,
 	struct vm_area_struct *vma)
 {
-	int offset, size, err;
+	struct snd_soc_pcm_runtime *rtd = substream->private_data;
+	struct snd_soc_dai *dai = rtd->cpu_dai;
+	int offset, size, err = 0;
 
-	/* TODO: we may need to check for underrun. */
-	vma->vm_flags |= VM_IO | VM_RESERVED;
-	vma->vm_page_prot = pgprot_writecombine(vma->vm_page_prot);
-	size = vma->vm_end - vma->vm_start;
-	offset = vma->vm_pgoff << PAGE_SHIFT;
-
-	err = io_remap_pfn_range(vma, vma->vm_start,
-			(ABE_DMEM_BASE_ADDRESS_MPU +
-			ABE_DMEM_BASE_OFFSET_PING_PONG + offset) >> PAGE_SHIFT,
-			size, vma->vm_page_prot);
+	switch (dai->id) {
+	case ABE_FRONTEND_DAI_LP_MEDIA:
+		/* TODO: we may need to check for underrun. */
+		vma->vm_flags |= VM_IO | VM_RESERVED;
+		vma->vm_page_prot = pgprot_noncached(vma->vm_page_prot);
+		size = vma->vm_end - vma->vm_start;
+		offset = vma->vm_pgoff << PAGE_SHIFT;
+
+		err = io_remap_pfn_range(vma, vma->vm_start,
+				(ABE_DMEM_BASE_ADDRESS_MPU +
+				ABE_DMEM_BASE_OFFSET_PING_PONG +
+				offset) >> PAGE_SHIFT,
+				size, vma->vm_page_prot);
+		break;
+
+	default:
+		break;
+	}
 
 	if (err)
 		return -EAGAIN;
Index: kernel/sound/soc/omap/omap-abe-dsp.h
===================================================================
--- kernel.orig/sound/soc/omap/omap-abe-dsp.h	2011-03-28 12:30:01.117144725 +0900
+++ kernel/sound/soc/omap/omap-abe-dsp.h	2011-03-28 15:57:27.708382549 +0900
@@ -24,10 +24,10 @@
 #ifndef __OMAP_ABE_DSP_H__
 #define __OMAP_ABE_DSP_H__
 
-#define ABE_DMEM_BASE_OFFSET_MPU	0x80000		/* Ping start offset of DMEM */
-#define ABE_DMEM_BASE_ADDRESS_MPU	0x49080000L	/* Pong start offset of DMEM */
-#define ABE_DMEM_BASE_OFFSET_PING_PONG	0x4000		/* Ping pong buffer DMEM offset */
-#define ABE_VM_AESS_OFFSET		0x400000	/* AESS VM offset for DMEM */
+/* Pong start offset of DMEM */
+#define ABE_DMEM_BASE_ADDRESS_MPU	0x40180000L
+/* Ping pong buffer DMEM offset */
+#define ABE_DMEM_BASE_OFFSET_PING_PONG	0x4000
 
 void abe_dsp_mcpdm_shutdown(void);
 void abe_dsp_pm_get(void);
